File: iosApp/iosApp/Network/JWTDecoder.swift
import Foundation
import OSLog

struct UserData: Codable {
  let status: String
  let firstname: String
  let lastname: String
  let email: String
  let country: String
  let locality: String

  var commonName: String {
    return "\(firstname) \(lastname)"
  }
}

class JWTDecoder {
  static private var logger = Logger.category("JWTDecoder")

  static func decodePayload<T: Codable>(token: String, as type: T.Type) -> T? {
    let parts = token.components(separatedBy: ".")
    guard parts.count == 3 else {
      return nil
    }

    var payload64 = parts[1]

    //INFO: format Base64Url -> Base64
    payload64 = payload64.replacingOccurrences(of: "-", with: "+")
      .replacingOccurrences(of: "-", with: "/")

    //INFO: Pad the payload
    while payload64.count % 4 != 0 {
      payload64 += "="
    }

    guard let payloadData = Data(base64Encoded: payload64) else { return nil }

    do {
      return try JSONDecoder().decode(T.self, from: payloadData)
    } catch {
      logger.error("Error during JWT data parsing: \(error.localizedDescription)")
      return nil
    }
  }
}

File: composeApp/src/androidMain/kotlin/cz/project/ewallet/Network/JWTDecoder.kt
package cz.project.ewallet

import android.util.Base64
import android.util.Log
import org.json.JSONObject

data class UserData(
        val status: String,
        val firstname: String,
        val lastname: String,
        val email: String,
        val country: String,
        val locality: String
) {
        val commonName: String
                get() = "$firstname $lastname"
}

object JWTDecoder {
        fun decodePayload(token: String): UserData? {
                try {
                        val parts = token.split(".")
                        if (parts.size != 3) return null

                        val payloadBase64 = parts[1]
                        val decodedBytes = Base64.decode(payloadBase64, Base64.URL_SAFE)
                        val decodedString = String(decodedBytes, Charsets.UTF_8)

                        val json = JSONObject(decodedString)

                        return UserData(
                                status = json.optString("status", "unknown"),
                                firstname = json.optString("firstname", "Neznámé"),
                                lastname = json.optString("lastname", "Jméno"),
                                email = json.optString("email", "no@mail.com"),
                                country = json.optString("country", "CZ"),
                                locality = json.optString("locality", "Neznámá lokalita")
                        )
                } catch (e: Exception) {
                        Log.e("JWTDecoder", "Error during token parsing: ${e.message}")
                        return null
                }
        }
}
